src/main/drivers/motor.c

```diff
@@ -0,0 +1,253 @@
+            motorDevice->vTable.write(i, values[i]);
+        }
+        motorDevice->vTable.updateComplete();
+    }
+#endif
+}
+
+int motorCount(void)
+{
+    return motorDevice->count;
+}
+
+// This is not motor generic anymore; should be moved to analog pwm module
+static void analogInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow) {
+    if (featureIsEnabled(FEATURE_3D)) {
+        float outputLimitOffset = (flight3DConfig()->limit3d_high - flight3DConfig()->limit3d_low) * (1 - outputLimit) / 2;
+        *disarm = flight3DConfig()->neutral3d;
+        *outputLow = flight3DConfig()->limit3d_low + outputLimitOffset;
+        *outputHigh = flight3DConfig()->limit3d_high - outputLimitOffset;
+        *deadbandMotor3dHigh = flight3DConfig()->deadband3d_high;
+        *deadbandMotor3dLow = flight3DConfig()->deadband3d_low;
+    } else {
+        *disarm = motorConfig()->mincommand;
+        *outputLow = motorConfig()->minthrottle;
+        *outputHigh = motorConfig()->maxthrottle - ((motorConfig()->maxthrottle - motorConfig()->minthrottle) * (1 - outputLimit));
+    }
+}
+
+// End point initialization is called from mixerInit before motorDevInit; can't use vtable...
+void motorInitEndpoints(float outputLimit, float *outputLow, float *outputHigh, float *disarm, float *deadbandMotor3dHigh, float *deadbandMotor3dLow)
+{
+    switch (motorConfig()->dev.motorPwmProtocol) {
+#ifdef USE_DSHOT
+    case PWM_TYPE_PROSHOT1000:
+    case PWM_TYPE_DSHOT1200:
+    case PWM_TYPE_DSHOT600:
+    case PWM_TYPE_DSHOT300:
+    case PWM_TYPE_DSHOT150:
+        dshotInitEndpoints(outputLimit, outputLow, outputHigh, disarm, deadbandMotor3dHigh, deadbandMotor3dLow);
+        break;
+#endif
+    default:
+        analogInitEndpoints(outputLimit, outputLow, outputHigh, disarm, deadbandMotor3dHigh, deadbandMotor3dLow);
+        break;
+    }
+}
+
+float motorConvertFromExternal(uint16_t externalValue)
+{
+    return motorDevice->vTable.convertExternalToMotor(externalValue);
+}
+
+uint16_t motorConvertToExternal(float motorValue)
+{
+    return motorDevice->vTable.convertMotorToExternal(motorValue);
+}
+
+static bool isDshot = false; // XXX Should go somewhere else
+
+static bool motorEnableNull(void)
+{
+    return false;
+}
+
+static void motorDisableNull(void)
+{
+}
+
+bool motorUpdateStartNull(void)
+{
+    return false;
+}
+
+void motorWriteNull(uint8_t index, float value)
+{
+    UNUSED(index);
+    UNUSED(value);
+}
+
+static void motorWriteIntNull(uint8_t index, uint16_t value)
+{
+    UNUSED(index);
+    UNUSED(value);
+}
+
+void motorUpdateCompleteNull(void)
+{
+}
+
+static void motorShutdownNull(void)
+{
+}
+
+static float motorConvertFromExternalNull(uint16_t value)
+{
+    UNUSED(value);
+    return(0.0f);
+}
+
+static uint16_t motorConvertToExternalNull(float value)
+{
+    UNUSED(value);
+    return 0;
+}
+
+static motorVTable_t motorNullVTable = {
```

Comment:
maybe `const` to put it in ROM?

Reply:
Yes! That's the whole purpose of copying the table at initialization.